api/v1/storageprofile_types.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2020 Red Hat OpenShift Container Storage.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StorageProfileSpec defines the desired state of StorageProfile
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.name) || has(self.name)", message="name is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.deviceClass) || has(self.deviceClass)", message="deviceClass is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.sharedFilesystemConfiguration) || has(self.sharedFilesystemConfiguration)", message="sharedFilesystemConfiguration is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.blockPoolConfiguration) || has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required once set"
+type StorageProfileSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Name is immutable"
+	// +kubebuilder:validation:MaxLength=512
+	// Name of the storage profile.
+	Name string `json:"name"`
```

Comment:
Every k8s resource already manages a name and namespace in its metadata. It is unnecessary to manage a name ourselves

Reply:
@nb-ohad Question is is that said `name` immutable by default? Think not. Then again maybe renaming SPs can be permitted, and it's the rest of the fields' immutability that should be enforced.

Reply:
@raaizik `Metadata.Name` is part of a resource identifier and has special meaning for k8s. You cannot change it. As @leelavg mentioned. If you change it and try to save the resource you create a new resource, not update the old one. Also it one of the fields you can select a resource by, where the name in spec (the one you added) does not have that property